src/bootstrap/builder.rs

```diff
@@ -832,7 +832,40 @@ impl<'a> Builder<'a> {
                 cargo.args(s.split_whitespace());
             }
             rustflags.env("RUSTFLAGS_BOOTSTRAP");
-            rustflags.arg("--cfg=bootstrap");
+            if cmd == "clippy" {
+                // clippy overwrites any sysroot we pass on the command line.
+                // Tell it to use the appropriate sysroot instead.
+                // NOTE: this can't be fixed in clippy because we explicitly don't set `RUSTC`,
+                // so it has no way of knowing the sysroot.
+                rustflags.arg("--sysroot");
+                rustflags.arg(
+                    self.sysroot(compiler)
+                        .as_os_str()
+                        .to_str()
+                        .expect("sysroot must be valid UTF-8"),
+                );
+                // Only run clippy on a very limited subset of crates (in particular, not build scripts).
+                cargo.arg("-Zunstable-options");
+                // Explicitly does *not* set `--cfg=bootstrap`, since we're using a nightly clippy.
+                let host_version = Command::new("rustc").arg("--version").output().map_err(|_| ());
+                if let Err(_) = host_version.and_then(|output| {
+                    if output.status.success()
+                        && t!(std::str::from_utf8(&output.stdout)).contains("nightly")
+                    {
+                        Ok(output)
+                    } else {
+                        Err(())
+                    }
+                }) {
+                    eprintln!(
+                        "error: `x.py clippy` requires a nightly host `rustc` toolchain with the `clippy` component"
```

Comment:
This seems like the wrong recommendation. Nightly clippy is not guaranteed to work; beta clippy is much more likely to.

Reply:
Nightly Clippy is always available, since we use `git subtree` for Clippy. It may have a few more FPs, but generally it should work. Or am I misunderstanding something?

Reply:
Nightly clippy is always available but may not work on the rustc tree -- it's usually outdated by several PRs (at most points in the day) which means that there's a pretty high likelihood of it not understanding some library feature or failing on stabilization etc. Beta clippy, like beta rustc, is something we're much more prepared for.

Reply:
I was not able to get this to work with a beta clippy.

```
error[E0522]: definition of an unknown language item: `array`
   --> library/core/src/array/mod.rs:379:1
    |
379 | #[lang = "array"]
    | ^^^^^^^^^^^^^^^^^ definition of unknown language item `array`
... various other E0522 errors ...
error[E0118]: no base type found for inherent implementation
   --> library/core/src/array/mod.rs:380:25
    |
380 | impl<T, const N: usize> [T; N] {
    |                         ^^^^^^ impl requires a base type
error: aborting due to 20 previous errors; 53 warnings emitted
```
This is with `cfg=bootstrap`. I think it's related to not setting `core` somehow.

Reply:
So, I am okay with us merging fixes to nightly clippy working, but I am unwilling to actually actively *recommend* nightly. I think that RUSTC getting set just needs to get fixed.

Reply:
I tried setting RUSTC to no effect, I think because there's no CLIPPY analogue to RUSTC or CLIPPY.

I'm not sure exactly what needs to happen to get beta clippy to work. Why does `x.py check` work with beta rustc when `x.py clippy` doesn't? I don't know what's going wrong.

Reply:
This started working after the beta bump, although I'm not sure why ... maybe I had an old version of beta or something? Anyway, I tested with both nightly and beta and it seems to be working.